indexing-service/src/main/java/io/druid/indexing/common/task/ArchiveTask.java

```diff
@@ -91,12 +98,15 @@ public TaskStatus run(TaskToolbox toolbox) throws Exception
       log.info("OK to archive segment: %s", unusedSegment.getIdentifier());
     }
 
-    // Move segments
-    for (DataSegment segment : unusedSegments) {
-      final DataSegment archivedSegment = toolbox.getDataSegmentArchiver().archive(segment);
-      toolbox.getTaskActionClient().submit(new SegmentMetadataUpdateAction(ImmutableSet.of(archivedSegment)));
+    if (toolbox.getTaskActionClient().submit(new LockUpgradeAction(getInterval()))) {
+      // Move segments
+      for (DataSegment segment : unusedSegments) {
+        final DataSegment archivedSegment = toolbox.getDataSegmentArchiver().archive(segment);
```

Comment:
After thinking more about it, I realized there can be problems if the lock is downgraded in critical section. Lets say for example an Archive Task was running in critical section and it downgraded the lock after archiving N segments. A realtime task comes in and revokes the Archive Task lock. The next time Archive Task tries to upgrade the lock (potentially after realtime task releases its lock) to archive the remaining segments, the upgrade will fail as there will be no lock to upgrade and thus task will fail without archiving all the segments, only some of them will be archived. Is this a problem in general for tasks having partially done their job ? 
If yes, one way to solve this problem is to create a `MayBeAcquireAndUpgrade` action which will first try to acquire lock (retry until it gets it) and then upgrade it as same time. This action will be used in critical section while upgrading the lock again to continue the critical section job.